projet/phase4/HK.jl

```diff
@@ -0,0 +1,118 @@
+"""Implémente l'algorithme de Held & Karp sur un graphe."""
+function HK(graphe_input::AbstractGraph{T}, iteration::Int, index_racine::Int) where T
+    # Initialisation des variables nécessaires
+    # graphe initial dont le poids des arêtes va être modifié
+    graphe = deepcopy(graphe_input)
+    nodes_graphe = getNodes(graphe)
+    edges_graphe = getEdges(graphe)
+    edges_true_weight = deepcopy(edges_graphe)
+
+    mst = initGraphPrim(graphe)
+    # données pour les itération de HK
+    n = length(nodes_graphe)
+    pi = zeros(n)
+    d = ones(n)
+    t = 1
+    k = 0
+    W = 0
+    sum = 0
+
+
+
+    # Construction du tableau de noeud et d'arêtes pour la construction de l'arbre de recouvrement minimal
+    nodes_mst = copy(nodes_graphe)
+    racine = nodes_mst[index_racine]
+    deleteat!(nodes_mst, index_racine)
+
+    edges_mst = copy(edges_graphe)
+    index_edges = findall(x -> isequal(x.node1, racine) || isequal(x.node2, racine), edges_mst)
+    edges_racine = edges_mst[index_edges]
+    deleteat!(edges_mst, index_edges)
+
+    edge1, edge2 = edges_min!(edges_racine)
+    # racine pour construire Prim devient l'extremité de la première arête
+    racine_prim = isequal(getNode1(edge1), racine) ? getNode2(edge1) : getNode1(edge1)
+    index_racine_prim = findfirst(x -> isequal(racine_prim, x), nodes_graphe)
+    lastNode = isequal(getNode1(edge2), racine) ? getNode2(edge2) : getNode1(edge2)
+    index_lastNode = findfirst(x -> isequal(lastNode, x), nodes_graphe)
+
+    while( k < iteration && d != zeros(n))
+        k += 1
```

Comment:
Vous pouvez rajouter un log à chaque itération (ou bien toutes les `q` itérations, où `q` est un paramètre).
Par exemple :
```
#Iter       C*          W
    0     1000       0.00
   10      983      57.23
   20      592     124.98
```